packages/prover/src/proof_provider/proof_provider.ts

```diff
@@ -0,0 +1,200 @@
+import {numberToHex} from "../utils/conversion.js";
+import {OrderedMap} from "./ordered_map.js";
+
+interface RootProviderOptions extends RootProviderInitOptions {
+  transport: LightClientRestTransport;
+  api: Api;
+  config: ChainForkConfig;
+}
+
+export class ProofProvider {
+  private payloads: OrderedMap<string, allForks.ExecutionPayload> = new OrderedMap();
+  private finalizedPayloads: OrderedMap<string, allForks.ExecutionPayload> = new OrderedMap();
+  private readyPromise: Promise<void>;
+
+  // Map to match CL slot to EL block number
+  private slotsMap: {[slot: number]: string} = {};
+
+  lightClient?: Lightclient;
+
+  constructor(private options: RootProviderOptions) {
+    this.readyPromise = this.sync(options.checkpoint);
+  }
+
+  async waitToBeReady(): Promise<void> {
+    return this.readyPromise;
+  }
+
+  static buildWithRestApi(urls: string[], opts: RootProviderInitOptions): ProofProvider {
+    const config = createChainForkConfig(networksChainConfig[opts.network]);
+    const api = getClient({urls}, {config});
+    const transport = new LightClientRestTransport(api);
+
+    return new ProofProvider({
+      ...opts,
+      config,
+      api,
+      transport,
+    });
+  }
+
+  private async sync(checkpoint?: string): Promise<void> {
+    if (this.lightClient !== undefined) {
+      throw Error("Light client already initialized and syncing.");
+    }
+
+    const {api, config, transport} = this.options;
+    const checkpointRoot = await getSyncCheckpoint(api, checkpoint);
+    const genesisData = await getGenesisData(api);
+
+    this.lightClient = await Lightclient.initializeFromCheckpointRoot({
+      checkpointRoot,
+      config,
+      transport,
+      genesisData,
+    });
+
+    assertLightClient(this.lightClient);
+    this.lightClient.start();
+    this.registerEvents();
+
+    const headSlot = this.lightClient.getHead().beacon.slot;
+    await this.lightClient?.sync(
+      computeSyncPeriodAtSlot(headSlot - MAX_REQUEST_LIGHT_CLIENT_UPDATES),
+      computeSyncPeriodAtSlot(headSlot)
+    );
+
+    // Load the payloads from the CL
+    const startSlot = headSlot;
+    const endSlot = headSlot - MAX_PAYLOAD_HISTORY;
+    const payloads = await getExecutionPayloads(this.options.api, startSlot, endSlot);
+    for (const [slot, payload] of Object.entries(payloads)) {
+      const blockNumber = numberToHex(payload.blockNumber);
+      this.payloads.set(blockNumber, payload);
+      this.slotsMap[parseInt(slot)] = blockNumber;
+    }
+
+    // Load the finalized payload from the CL
+    const finalizedSlot = this.lightClient.getFinalized().beacon.slot;
+    const finalizedPayload = await getExecutionPayloads(this.options.api, finalizedSlot, finalizedSlot);
+    const finalizedBlockNumber = numberToHex(finalizedPayload[finalizedSlot].blockNumber);
+    this.finalizedPayloads.set(finalizedBlockNumber, finalizedPayload[finalizedSlot]);
```

Comment:
again why do we need this mapping, once a slot is finalized, all numbers corresponding to finalized's block number and before are finalized, why do we need this mapping